### PR TITLE
feat: port of DETR

### PR DESCRIPTION
**Pull request by liamcli**
## Description
Port of DETR: https://github.com/facebookresearch/detr/tree/master
[DET-4403]

## Test Plan
Confirmed performance matches reported logs for 150 epoch training:
https://gist.github.com/szagoruyko/b4c3b2c3627294fc369b899987385a3f

## Commentary 
We can try to further improve performance over the original library by:
1. Substituting in EfficientNet back bone in place of ResNet backbone.
2. Use the evolved transformer architecture in place of the standard transformer architecture.

[DET-4403]: https://determinedai.atlassian.net/browse/DET-4403

### CODE REVIEW

File: examples/computer_vision/detr/README.md [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+DETR drastically simplifies the object detection pipeline by doing away with complex postprocessing steps required by most other architectures to extract bounding boxes.  
+In place of proposals, anchors, and window centers, DETR directly optimizes a set prediction loss to learn where to focus.  
+
+This implementation largely uses the original implementation provided in [this repo](https://github.com/facebookresearch/detr).  With this Determined implementation, you can easily run DETR with distributed training and hyperparameter search by modifying a few flags in the [experiment config](distributed.yaml).
+
+## Files
+This example is structure as follows:
+* [model_def.py](model_def.py): initializes the data, optimizer, and model and specifies the training and evaluation steps.
+* [model.py](model.py): modifies the original [SetCriterion function](https://github.com/facebookresearch/detr/blob/master/models/detr.py#L83) to use horovod allreduce to sync number of bounding boxes.
+* [data.py](data.py): creates a CocoDetection dataset to work with Google Cloud Storage.
```
Comment (naren-determined) [inline examples/computer_vision/detr/README.md:12]: Can you create a subsection in `## Files` called `### Configuration Files`, the latter of which has all the associated .yaml?

File: examples/computer_vision/detr/README.md [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+This example is structure as follows:
+* [model_def.py](model_def.py): initializes the data, optimizer, and model and specifies the training and evaluation steps.
+* [model.py](model.py): modifies the original [SetCriterion function](https://github.com/facebookresearch/detr/blob/master/models/detr.py#L83) to use horovod allreduce to sync number of bounding boxes.
+* [data.py](data.py): creates a CocoDetection dataset to work with Google Cloud Storage. 
+* [distributed.yaml](distributed.yaml): distributed training experiment to run with COCO dataset on Google Cloud Storage.
+* [const_fake.yaml](const_fake.yaml): single-GPU experiment to run with fake data.
+
+
+## Running the example
+If you have the COCO 2017 dataset available in a Google Cloud Storage bucket, you can modify the `bucket_name` field in [distributed.yaml](distributed.yaml) to point to your location.  Otherwise, you can run this example with fake data using the [const_fake.yaml](const_fake.yaml) config.  Submit the experiment to your Determined cluster by running
```
Comment (naren-determined) [inline examples/computer_vision/detr/README.md:18]: Any chance we can have a `self.data_downloaded` flag set in the `__init__` and download COCO from a URL in Determined way that users don't have to worry about the whole Google Cloud Storage upload if they want to run with real COCO data?

File: examples/computer_vision/detr/const_fake.yaml [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    epochs: 1
+records_per_epoch: 117264
+searcher:
+  name: single
+  metric: mAP
+  smaller_is_better: false
+  max_length:
+      epochs: 150
+resources:
+    slots_per_trial: 1
```
Comment (naren-determined) [inline examples/computer_vision/detr/const_fake.yaml:58]: This line is somewhat unnecessary since the default value is 1. It is implied that const is only single GPU.

File: examples/computer_vision/detr/data.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        return img_str
+
+
+class FakeBackend:
+    def __init__(self):
+        self.data = None
+
+    def get(self, filepath):
+        if self.data is None:
+            with open("imgs/train_curves.png", "rb") as f:
```
Comment (naren-determined) [inline examples/computer_vision/detr/data.py:50]: I know this is generating fake data from the training curve, but can we maybe put a comment here just to be clear on that?

File: examples/computer_vision/detr/model.py [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+    )
+    if args.masks:
+        model = DETRsegm(model, freeze_detr=(args.frozen_weights is not None))
+    matcher = build_matcher(args)
+    weight_dict = {"loss_ce": 1, "loss_bbox": args.bbox_loss_coef}
+    weight_dict["loss_giou"] = args.giou_loss_coef
+    if args.masks:
+        weight_dict["loss_mask"] = args.mask_loss_coef
+        weight_dict["loss_dice"] = args.dice_loss_coef
+    # TODO this is a hack
```
Comment (naren-determined) [inline examples/computer_vision/detr/model.py:272]: Can you fix this TODO?

File: examples/computer_vision/detr/model_def.py [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+        self.hparams = AttrDict(self.context.get_hparams())
+
+        # Wrap the model.
+        model, self.criterion, self.postprocessors = build_model(
+            self.hparams, world_size=self.context.distributed.get_size()
+        )
+        self.model = self.context.wrap_model(model)
+
+        n_parameters = sum(p.numel() for p in model.parameters() if p.requires_grad)
+        print("number of params:", n_parameters)
```
Comment (naren-determined) [inline examples/computer_vision/detr/model_def.py:45]: Should we use `Logger.log` here instead of a print statement?

File: examples/computer_vision/detr/model_def.py [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+        )
+
+        self.clip_grads_fn = (
+            lambda x: torch.nn.utils.clip_grad_norm_(x, self.hparams.clip_max_norm)
+            if self.hparams.clip_max_norm > 0
+            else None
+        )
+
+        # Build datasets
+        self.dataset_train = build_dataset(image_set="train", args=self.hparams)
```
Comment (naren-determined) [inline examples/computer_vision/detr/model_def.py:83]: Can both dataset lines be moved to `build_training_data_loader` and `build_validation_data_loader` respectively?

File: examples/computer_vision/detr/model_def.py [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+        self.dataset_val = build_dataset(image_set="val", args=self.hparams)
+
+        # Set up evaluator
+        self.base_ds = get_coco_api_from_dataset(self.dataset_val)
+        self.iou_types = tuple(
```
Comment (naren-determined) [inline examples/computer_vision/detr/model_def.py:88]: As this is only used in `evaluate_full_dataset`, this doesn't need to be in the __init__

File: examples/computer_vision/detr/model_def.py [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+                "lr": self.hparams.lr_backbone,
+            },
+        ]
+        self.optimizer = self.context.wrap_optimizer(
+            torch.optim.AdamW(
+                param_dicts, lr=self.hparams.lr, weight_decay=self.hparams.weight_decay
+            )
+        )
+
+        # Wrap the LR scheduler.
```
Comment (naren-determined) [inline examples/computer_vision/detr/model_def.py:70]: Can we put this in https://docs.determined.ai/latest/reference/api/pytorch.html?highlight=lr_scheduler#determined.pytorch.PyTorchTrialContext.wrap_lr_scheduler just for easy of readability and a bit more structured?

File: examples/computer_vision/detr/model_def.py [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+        self.lr_scheduler = self.context.wrap_lr_scheduler(
+            torch.optim.lr_scheduler.StepLR(self.optimizer, self.hparams.lr_drop),
+            step_mode=LRScheduler.StepMode.STEP_EVERY_EPOCH,
+        )
+
+        self.clip_grads_fn = (
```
Comment (naren-determined) [inline examples/computer_vision/detr/model_def.py:76]: I don't see this function used anywhere else. Can we delete it if that's the case?

File: examples/computer_vision/detr/data.py [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+        if self._transforms is not None:
+            img, target = self._transforms(img, target)
+        return img, target
+
+
+def build_dataset(image_set, args):
+    root = args.bucket_name
+    mode = "instances"
+    PATHS = {
+        "train": (f"{root}/train2017", f"/tmp/{mode}_train2017.json"),
```
Comment (naren-determined) [inline examples/computer_vision/detr/data.py:89]: Don't these data paths need to be cached per rank like `f"/tmp/data-rank{self.context.distributed.get_rank()}"`?

**Review by naren-determined — CHANGES_REQUESTED** [review]
You have a `detr/log_25935349.txt` file being submitted as part of this PR. Is that intentional? 
Overall, my main feedback is around cleaning up the `__init__` function in model_def.py. Additionally, can you follow these instructions on adding basic tests and stuff to documentation/READMEs: https://docs.google.com/document/d/1G5pKRllMnNXnLQKYlHGMjOfFz66OHUcGRVnSLVMs8WQ/edit?

File: examples/computer_vision/detr/model.py (hunk 5 above)
Comment (liamcli) [reply examples/computer_vision/detr/model.py:272]: this TODO is from the original repo.  i removed and added a header comment to explain that this file is copied from the original repo with slight modifications for dtrain.

File: examples/computer_vision/detr/README.md (hunk 2 above)
Comment (liamcli) [reply examples/computer_vision/detr/README.md]: added a local backend for data that will download coco from source if directories do not already exist.  coco is fairly large so this will take >5mins.

File: examples/computer_vision/detr/data.py (hunk 11 above)
Comment (liamcli) [reply examples/computer_vision/detr/data.py:110]: i've also added logic in model_def.py to only download on local_rank 1 slot.

File: examples/computer_vision/detr/model_def.py (hunk 6 above)
Comment (liamcli) [reply examples/computer_vision/detr/model_def.py:66]: given there are only two print statements in this example, i think okay to not use logging module.

File: examples/computer_vision/detr/model_def.py (hunk 9 above)
Comment (naren-determined) [reply examples/computer_vision/detr/model_def.py]: We have a `wrap_lr_scheduler` function as part of the PyTorch API, so my suggestion was to take the 4 or 5 lines regarding LR scheduler in the `model_def.py`, move them out of the __init__ and into that function.

File: e2e_tests/tests/nightly/test_distributed.py [hunk 12]
```diff
@@ -94,3 +94,12 @@ def test_gan_mnist_pytorch_distributed() -> None:
     config = conf.set_max_length(config, {"batches": 200})
 
     exp.run_basic_test_with_temp_config(config, conf.gan_examples_path("gan_mnist_pytorch"), 1)
+
+
+@pytest.mark.distributed  # type: ignore
+def test_detr_coco_pytorch_distributed() -> None:
+    config = conf.load_config(conf.cv_examples_path("detr_coco_pytorch/const_fake.yaml"))
```
Comment (naren-determined) [inline e2e_tests/tests/nightly/test_distributed.py:101]: I think you should make a distributed_fake.yaml for this instead of manually setting slots to 2.

File: examples/computer_vision/detr_coco_pytorch/README.md [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+det experiment create <distributed_gcs.yaml> .
+```
+from the command line.
+
+## Results
+The training and validation curves generated by this experiment are shown below.  They closely match the [results for the 150 epoch learning rate schedule provided in the original repo](https://gist.github.com/szagoruyko/b4c3b2c3627294fc369b899987385a3f).
+
+![train_curves](imgs/train_curves.png)
+![val_curves](imgs/val_curves.png)
+
```
Comment (naren-determined) [inline examples/computer_vision/detr_coco_pytorch/README.md:34]: Nit: Remove extra lines.

File: examples/computer_vision/detr_coco_pytorch/data_utils.py [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+
+
+def download_file(url, output_dir):
+    local_filename = os.path.join(output_dir, url.split("/")[-1])
+    with requests.get(url, stream=True) as r:
+        r.raise_for_status()
+        with open(local_filename, "wb") as f:
+            for chunk in r.iter_content(chunk_size=8192):
+                # If you have chunk encoded response uncomment if
+                # and set chunk_size parameter to None.
```
Comment (naren-determined) [inline examples/computer_vision/detr_coco_pytorch/data_utils.py:16]: Would condense the two line explanation of comment into one just for readibility, and say uncomment "the line below"

File: examples/computer_vision/detr_coco_pytorch/model_def.py [hunk 15]
```diff
@@ -0,0 +1,219 @@
...
+            },
+        ]
+        self.optimizer = self.context.wrap_optimizer(
+            torch.optim.AdamW(
+                param_dicts, lr=self.hparams.lr, weight_decay=self.hparams.weight_decay
+            )
+        )
+
+        # Wrap the LR scheduler.
+        self.lr_scheduler = self.context.wrap_lr_scheduler(
```
Comment (naren-determined) [inline examples/computer_vision/detr_coco_pytorch/model_def.py:92]: Re: convo on LR scheduler function

File: examples/computer_vision/detr_coco_pytorch/startup-hook.sh [hunk 16]
```diff
@@ -0,0 +1,12 @@
+apt-get update
```
Comment (naren-determined) [inline examples/computer_vision/detr_coco_pytorch/startup-hook.sh:1]: Why do we need this in the startup hook?

Comment (liamcli) [reply examples/computer_vision/detr_coco_pytorch/startup-hook.sh:1]: for some reason this is needed to install unzip.

**Review by naren-determined — APPROVED** [review]
🚢
